### ISSUE TITLE
same_subject constraint seems to have limited utility

### ISSUE THREAD

**Issue by joshbax189** [decentralized-identity/presentation-exchange#207]
Since an input descriptor may be satisfied with only a single input, the `same_subject` constraint can only have effect on multi-subject credentials (or presentations). 

JWT credentials and presentations do not support multiple subjects. (A note in the W3C VC spec says only that they "might" in future). Issuers who want their credentials to be available in multiple formats would default to using a single subject. So, considering the available formats, the multi-subject case is the exception.

Under the assumption that most credentials would have a single subject, this constraint would hardly ever apply.

Possible solution:
Move this constraint to submission requirements so it can apply to multiple inputs. For example,
```
{
  "name": "Submission of educational transcripts",
  "purpose": "We need your complete educational transcripts to process your application",
  "rule": "all",
  "from": "A",
  "same_subject": "required"
}
```
would mean all claims used to satisfy the requirement must have the same subject (same for a "pick" rule).

The processing algorithm would have to change:
- input evaluation should return ALL claims that satisfy the descriptor
- checking a requirement with a `same_subject` constraint would involve grouping claims by subject, then checking if the required number of inputs were satisfied
- checking a requirement without such a constraint would default to taking the first claim per input descriptor

Although it seems complex, consider how a VC Holder would go about satisfying the following input descriptor if they just had the `drivers_license` and `postal_address` VCs.
```
{
  "id": "license_and_address",
  "schema": [
    {
      "uri": "https://example.com/drivers_license.json",
      "required": true
    },
    {
      "uri": "https://example.com/postal_address.json",
      "required": true
    }
  ],
  "constraints": {
    "same_subject": {
      "field_id": [ "credential_subject" ],
      "directive": "required"
    },
    "fields": [
      "id": "credential_subject",
      "purpose": "Extract credential subject nodes from an LDP presentation",
      "path": [ "$.verifiableCredential[*].credentialSubject" ]
    ]
  }
}
```

**Comment by brentzundel** [comment]
the `field_id` property of a `same_subject` constraint, lists all of the `field.id`s of all the claims that should have the same subject. The spec explicitly states that such `field.id`s may come from other input descriptors.

This allows a verifier to use the `same_subject` constraint to require that the `first_name` claim from one JWT has the same subject as an `account_balance` claim from another JWT. 

Making the `same_subject` constraint less fine-grained would make certain use cases impossible, particularly in the case where a credential has multiple subjects, so I definitely don't agree that it should be moved to the `submission_requirements`.

It seems we disagree about the likely prevalence of multi-subject credentials. You assert that most credentials will have a single subject. I'm not sure that will be the case. Regardless, since `same_subject` may be used across `input_descriptors` and therefore across credentials. It seems as though this constraint would be useful even if your assertion ends up being correct.

I do not believe this issue warrants any changes to the specification.
Perhaps I've misunderstood your issue. If I have, I apologize, and ask that you clarify.

**Comment by joshbax189** [comment]
Thanks for your reply! (... and thanks in advance for reading a rather long reply, TL;DR is at the end ;) )

You're right: I didn't notice that field ids could come from other input descriptors. My assumption was that only the field ids defined in the descriptor could be used (that also makes the point about prevalence of multi-subject credentials moot, so let's leave that aside).

I still believe that the same-subject constraint should be moved to the submission requirements object, mainly for developer/user sanity.

Consider the difficulty of encoding a same-subject constraint in a presentation definition for [Australian Identity Requirements](https://www.passports.gov.au/getting-passport-how-it-works/documents-you-need/identity). It has 20+ optional inputs, each of which should have matching subjects.
This would have input descriptors
```
[ { 
  "id": "document_1",
  ...
  "fields": [{
     "id": "subject_field_1",
     ... }]
  },
  ...,
  {
    "id": "document_20",
    ...
    "fields": [{
       "id": "subject_field_20",
       ... }]
    }]
```
But then where does the same-subject constraint go?
* On every input descriptor? 
  Leads to 20x 20 element arrays (or 19 element arrays excluding the parent input's subject), or
* On a single input? 
  But which one? How will you know that the other 19 inputs are subject to that constraint?

If the same-subject constraint were in the submission requirements, then it would be obvious which input descriptors are constrained, where the constraint should be written, and it is only written once.

> Making the same_subject constraint less fine-grained would make certain use cases impossible

By less fine-grained I understand that you mean the constraint doesn't refer to fields? I think that's a useful simplification, because the credential subject should always be in a fixed location relative to the claim used to satisfy the input descriptor. That location can be deduced by looking at the `format` in the `descriptor_map`.

For multi-subject credentials, this is only a problem if the input descriptor specifically requires multiple subjects, like in my bad example above. If you were using a multi-subject credential to satisfy an input descriptor that requests a single subject claim then the correct claim subject can be deduced from the presentation submission.

Otherwise, my suggestion is not much changed if a `field_ids` array were included with the `same_subject` constraint in the submission requirements:
```
{
  "name": "Submission of educational transcripts",
  "purpose": "We need your complete educational transcripts to process your application",
  "rule": "all",
  "from": "A",
  "same_subject": {
    "directive": "required",
    "field_ids": [ "subject_field_1", ... ]
  }
}
``` 

In fact, including the `field_ids` array would allow for an interesting generalization of the same-subject constraint to a "same-value" constraint, as the fields could describe any claim within an input. 

(Side note: now that I think about it, doesn't the same reasoning apply to the existing `same_subject` constraint? Since the fields can be any claim, not only the claim/credential subject, then that constraint could be abused to be a "same value" constraint-- perhaps it should be renamed at least?)

TL;DR:
* It seems like the least ambiguous place to put the `same_subject` constraint would be the submissions requirements; perhaps it could be cut-and-pasted there?
* If you assume that input descriptors don't specifically request multi-subject credentials (like my bad example), then the credential/claim subject should be in a location defined by the `descriptor_map`, so the submission requirements form of `same_subject` could be simplified by removing `field_ids`.
* Or, `same_subject` could be generalized to `same_value`, since the `field_ids` can point to anything they like. (This observation applies to `same_subject` in it's current form too.)

**Comment by brentzundel** [comment]
I don't disagree that having the `same_subject` in the `submission_requirements` might make it more clear what the full set of `same_subject` constraints may be, but I'm opposed to moving them for a number of reasons:
1. `submission_requirements` are used to describe which logical combinations of groups of inputs are acceptable. Using them to also define constraints on specific inputs doesn't fit within that scope..
2. `input_descriptors` are required, but `submission_requirements` are not.
3. I don't think it makes sense to have a property at the `submission_requirements` level that points explicitly to the `input_descriptors` level. The structure of a `presentation_definition` at this point only has `input_descriptors` point up to different `submission_requirements`. Adding a bi-directional dependency feels wrong to me.
4. `same_subject` describes a constraint on a specific input, so it makes sense to me for it to be part of that `input_descriptors` `constraints` object.

I also don't think it makes sense to get rid of `field_id`s. This, combined with your third point leads me to think you may still have a slight misunderstanding about what `same_subject` is for.

The `same_subject` `field_id`s indicate which properties should be about the same subject. Those properties could be anything.
When using `same_subject`, the verifier isn't saying, "make sure these two subject properties have the same value," but rather, "make sure these two properties are about the same subject." i.e., as a verifier, I want to be sure that the first name property from the birth certificate credential and the GPA property from the transcript credential are about the same subject.

**Comment by brentzundel** [comment]
spec has been significantly re-factored since this issue was raised. I believe it has been addressed.

**Comment by csuwildcat** [comment]
Closing, has been addressed.
